examples/car_statem.erl

```diff
@@ -0,0 +1,314 @@
+         next_state/3]).
+
+
+%% -----------------------------------------------------------------------------
+%% Definitions
+%% -----------------------------------------------------------------------------
+
+
+-define(DISTANCE, 1500).
+
+
+%% -----------------------------------------------------------------------------
+%% Records
+%% -----------------------------------------------------------------------------
+
+
+-record(state,
+        {fuel  :: float(),
+         speed :: non_neg_integer()}).
+
+-record(test_state,
+        {fuel     :: float(),
+         speed    :: non_neg_integer(),
+         distance :: float(),
+         burnt    :: float()}).
+
+
+%% -----------------------------------------------------------------------------
+%% Common Imports
+%% -----------------------------------------------------------------------------
+
+
+-include("car.inc").
+
+
+%% -----------------------------------------------------------------------------
+%% API
+%% -----------------------------------------------------------------------------
+
+
+start_link() ->
+  gen_server:start_link({local, ?NAME}, ?MODULE, [], []).
+
+stop() ->
+  gen_server:stop(?NAME).
+
+accelerate(Value) ->
+  gen_server:call(?NAME, {accelerate, Value}).
+
+brake(Value) ->
+  gen_server:call(?NAME, {brake, Value}).
+
+travel(Distance) ->
+  gen_server:call(?NAME, {travel, Distance}).
+
+refuel(Amount) ->
+  gen_server:call(?NAME, {refuel, Amount}).
+
+
+%% -----------------------------------------------------------------------------
+%% gen_server callbacks
+%% -----------------------------------------------------------------------------
+
+
+init([]) ->
+  {ok, #state{fuel = ?MAX_FUEL, speed = 0}}.
+
+handle_call({accelerate, Value}, _From, S) ->
+  #state{fuel = Fuel, speed = Speed} = S,
+  Calc = acceleration_calculations(Speed, Value, Fuel),
+  {Distance, Acceleration, Burnt} = Calc,
+  NS = S#state{fuel = Fuel - Burnt, speed = Speed + Acceleration},
+  {reply, {Distance, Burnt}, NS};
+
+handle_call({brake, Value}, _From, S) ->
+  #state{fuel = Fuel, speed = Speed} = S,
+  Calc = acceleration_calculations(Speed, -Value, Fuel),
+  {Distance, Deceleration, Burnt} = Calc,
+  NS = S#state{fuel = Fuel - Burnt, speed = Speed + Deceleration},
+  {reply, {Distance, Burnt}, NS};
+
+handle_call({travel, Distance}, _From, S) ->
+  #state{fuel = Fuel, speed = Speed} = S,
+  {RealDistance, Burnt} = travel_calculations(Distance, Speed, Fuel),
+  NS = S#state{fuel = Fuel - Burnt},
+  {reply, {RealDistance, Burnt}, NS};
+
+handle_call({refuel, Value}, _From, S) ->
+  #state{fuel = Fuel, speed = Speed} = S,
+  Calc = acceleration_calculations(Speed, -Speed, Fuel),
+  {Distance, _, Burnt} = Calc,
+  NS = S#state{fuel = min(?MAX_FUEL, Fuel - Burnt + Value), speed = 0},
+  {reply, {Distance, Burnt}, NS}.
+
+handle_cast(_Msg, S) ->
+  {noreply, S}.
+
+handle_info(_Msg, S) ->
+  {noreply, S}.
+
+terminate(_Reason, _S) ->
+  {ok}.
```

Comment:
This reads weird.  A simple `ok` will do.

Reply:
Oops :)